File: 2023-j5-v3.c
#include <iostream>
#include <vector>
#include <stack>
#include <unordered_map>
#include <set>

using namespace std;

// 判断点是否共线的函数
bool are_points_collinear(const vector<pair<int, int>>& points) {
    // 如果点的数量小于3，它们自然在一条直线上
    if (static_cast<int>(points.size()) < 3) return true;

    // 获取前两个点的坐标
    int x0 = points[0].first, y0 = points[0].second;
    int x1 = points[1].first, y1 = points[1].second;
    int dx = x1 - x0;
    int dy = y1 - y0;

    // 计算其余点的斜率
    for (int i = 2; i < static_cast<int>(points.size()); i++) {
        int x = points[i].first, y = points[i].second;
        if (dx * (y - y0) != dy * (x - x0)) {
            return false;
        }
    }
    return true;
}

// 深度优先搜索栈实现
int dfs_stack(int start_x, int start_y, const vector<vector<string>>& grid, const vector<char>& word_list, int rows, int columns) {
    int result = 0;
    stack<vector<pair<int, int>>> s;
    s.push({{start_x, start_y}});

    while (!s.empty()) {
        vector<pair<int, int>> path = s.top();
        s.pop();
        int x = path.back().first;
        int y = path.back().second;

        if (path.size() == word_list.size()) {
            result++;
            continue;
        }

        char next_key = word_list[path.size()];

        vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}, {-1, 1}, {1, 1}, {-1, -1}, {1, -1}};

        for (auto& dir : directions) {
            int nx = x + dir.first;
            int ny = y + dir.second;

            // 检查是否越界并且下一个字母是否匹配
            if (nx >= 0 && nx < rows && ny >= 0 && ny < columns && grid[nx][ny] == string(1, next_key)) {
                int step = abs(nx - x) + abs(ny - y);

                // 如果路径长度为1，直接推进栈
                if (path.size() == 1) {
                    s.push(path);
                    s.top().emplace_back(nx, ny);
                }

                // 如果路径长度大于1，进行更多检查
                if (path.size() > 1) {
                    int lx = path[path.size() - 2].first;
                    int ly = path[path.size() - 2].second;
                    int px = 2 * x - lx;
                    int py = 2 * y - ly;
                    int step01 = abs(path[0].first - path[1].first) + abs(path[0].second - path[1].second);

                    if (make_pair(px, py) == make_pair(nx, ny)) {
                        s.push(path);
                        s.top().emplace_back(nx, ny);
                    } else if (step01 == step && are_points_collinear(path)) {
                        s.push(path);
                        s.top().emplace_back(nx, ny);
                    }
                }
            }
        }
    }
    return result;
}

int main() {
    // 读取输入
    string input_word;
    cin >> input_word;

    vector<char> word_list(input_word.begin(), input_word.end());

    int rows, columns;
    cin >> rows >> columns;

    vector<vector<string>> grid(rows, vector<string>(columns));

    // 读取网格
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < columns; ++j) {
            cin >> grid[i][j];
        }
    }

    // 查找与word_list第一个字符匹配的元素
    vector<pair<int, int>> leading_elements;
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < columns; ++j) {
            if (grid[i][j] == string(1, word_list[0])) {
                leading_elements.emplace_back(i, j);
            }
        }
    }

    // 统计符合条件的路径
    int total = 0;
    for (auto& l : leading_elements) {
        total += dfs_stack(l.first, l.second, grid, word_list, rows, columns);
    }

    // 输出结果
    cout << total << endl;
    return 0;
}

File: 2023-j5-v3.py

word_list = list(input())
rows = int(input())
columns = int(input())
grid = []
for i in range(rows):
    grid.append(input().split(' '))
path_dict = {key: 0 for key in word_list}
findings = set()

def are_points_collinear(points):
    if len(points) < 3:  # 如果点的数量小于3，它们自然在一条直线上
        return True

    # 计算第一个点和第二个点的斜率
    (x0, y0), (x1, y1) = points[0], points[1]
    dx = x1 - x0
    dy = y1 - y0

    for i in range(2, len(points)):
        x, y = points[i]
        # 对于每个后续的点，计算与第一个点的斜率
        if dx * (y - y0) != dy * (x - x0):
            return False  # 如果斜率不相同，返回False

    return True  # 所有斜率相同，返回True

def dfs_stack(start_x, start_y):
    result = 0
    stack = [[(start_x, start_y)]]
    while stack:
        path = stack.pop()
        x, y = path[-1]
        if len(path)== len(word_list):
            result+=1
            continue
        next_key= word_list[len(path)]

        for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, 1), (1, 1), (-1, -1), (1, -1)]:
            nx, ny = x + dx, y + dy
            if 0 <= nx < rows and 0 <= ny < columns and grid[nx][ny] == next_key:
                step = abs(nx - x) + abs(ny - y)
                turning = path[0][1]
                if len(path) == 1:
                    stack.append( path+ [(nx, ny)])
                if len(path) > 1:
                    step01=abs(path[0][1] - path[1][1])+abs(path[0][0] - path[1][0])
                    lx, ly = path[-2][0], path[-2][1]
                    px, py = 2 * x - lx, 2 * y - ly
                    if (px, py) == (nx, ny):
                        stack.append(path+ [(nx, ny)])
                    elif step01 == step and are_points_collinear(path):
                        stack.append(path+ [(nx, ny)])
    return result


leading_elements = []
for i in range(rows):
    for j in range(columns):
        if grid[i][j] == word_list[0]:
            leading_elements.append((i, j))
total = 0
for l in leading_elements:
    total += dfs_stack(l[0], l[1])

print(total)

